### ISSUE TITLE
Web mode as a meta mode

### ISSUE THREAD

**Issue by jmercouris** [atlas-engineer/nyxt#1851]
I noticed that many of our modes are based upon web-mode. However, web mode is a very large mode! I am thinking of the following, turn web-mode into a meta mode.

Toggling web-mode will actually toggle a set of modes (jump heading mode, search buffer mode, etc).

For example, we'll add a jump-heading mode that web-mode will toggle. What do you think? A logical way to decompose web-mode?

@Ambrevar

**Comment by jmercouris** [comment]
Definition: a meta mode is a mode that toggles modes.

**Comment by Ambrevar** [comment]
Why not, but first a question: is there a use case where one would want or need `jump-heading-mode` without `search-buffer-mode`?

In other words, does this decomposition make sense in terms of "mode units", from the user perspective?

Alternative suggestion: we could further split web-mode.lisp into multiple files, one for history, one for bindings, etc.

**Comment by jmercouris** [comment]
Perhaps it does not make sense... in any case, I want a way to make `web-mode` smaller and more focused. We can still bind things to disabled modes. Perhaps web-mode should just be a common set of bindings for `jump-heading` mode, etc.?

**Comment by aadcg** [comment]
I bumped into this from another perspective, while working on #2195.

I drew a silly graph from the way we structure our modules. The arrow "A -> B" defines "A depends on B". The 2 arrows I stroke out (emanating from "Help") are those I find illogical. I replaced those 2 with one connecting "Help" to "web-mode". 

"web-mode" is actually "web-mode commands".

![2022-04-22_10:13:24](https://user-images.githubusercontent.com/45483512/164676309-d7d9890d-3aed-4f1c-b0ab-e029741371d1.png)

**Comment by Ambrevar** [comment]
Yes, `help`-related code needs to be reorganized some bits.

Note that I did reorganize the asd modules in an ad-hoc fashion: it's better than before but  by no means fully "logical", it's just good enough as long as we have this illogical mutual dependencies in the code base (web-mode, help, etc.).  

In other words: no need to focus on the .asd, we should focus on the modes vs. the core instead.

**Comment by aadcg** [comment]
I see your point, and it does make sense.

I'm afraid I'm not qualified enough to give much advice, so I'll perhaps wait until a consensus is reached. 

Regarding `help`, as seen as a module, it could be broken down into two: `help` and `help-interface`. The former defines commands to inspect Nyxt, while the later (depending on web-mode commands such as `jump-to-heading`) defines the interface of what we today call `help-mode`. Concretely, this would means that `mode/help` wouldn't be part of "Core modes" anymore.

Again, this is my naive perspective.

**Comment by Ambrevar** [comment]
Here is my breakdown suggestion for `web-mode`:


- navigation-mode.  Merge the following:

  - copy/paste/ring management
	- undo/redo
	- scroll
	- search-buffer
	- zoom
	- jump-to-heading

- spell-check-mode

- history-mode

- hint-mode.  Files (maybe merge):

  - element-frame
  - element-hint

Nyxt geeky modes:

- summarize-mode

- autofill-mode

- bookmarklets-mode

To do:

- [ ] Move `focus-first-input-field` to `element-hint.lisp`?

Questions:

- `navigation-mode` only contains the basic features found in all browser.
  Except for `jump-to-heading`.  Move it to a separate mode then?

- Should we still have a `web` meta mode?
  If yes, then shall we name it differently, e.g. `document-mode`, since it's
  not just for the web but any markup document.

  If no, we set all the above modes as the default modes for web-buffers.  It's
  easy for the user to disable some of them if they want to.

Thoughts?

**Comment by Ambrevar** [comment]
Another thought: base-mode was historically needed to find global commands, but no longer need this rule.
So it's just a bunch of regular navigation keys...

So what about merging `navigation-mode` and `base-mode` (I like the name `navigation` more since it's more explicit).

Then we don't need `base-mode` / `navigation-mode` as a default mode everywhere, we just need it in `web-buffers`.

**Comment by aartaka** [comment]
Yes! I've been long irritated by the fact that some `base-mode` commands semantically belong to `web-mode` and vice versa!
